pkg/controllers/awsloadbalancercontroller/deployment.go

```diff
@@ -2,13 +2,222 @@ package awsloadbalancercontroller
+	appName         = "aws-load-balancer-operator"
+	appInstanceName = "app.kubernetes.io/instance"
+)
+
 func (r *AWSLoadBalancerControllerReconciler) ensureControllerDeployment(ctx context.Context, namespace string, image string, sa *corev1.ServiceAccount, controller *albo.AWSLoadBalancerController) (*appsv1.Deployment, error) {
-	return nil, nil
+	deploymentName := fmt.Sprintf("%s-%s", controllerResourcePrefix, controller.Name)
+
+	exists, current, err := r.currentControllerDeployment(ctx, deploymentName, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get existing deployment %s: %w", deploymentName, err)
+	}
+
+	desired := desiredDeployment(deploymentName, namespace, image, r.VPCID, r.ClusterName, controller, sa)
+	err = controllerutil.SetOwnerReference(controller, desired, r.Scheme)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set owner reference on deployment %s: %w", deploymentName, err)
+	}
+
+	if !exists {
+		err = r.createDeployment(ctx, desired)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create deployment %s: %w", deploymentName, err)
+		}
+		_, current, err = r.currentControllerDeployment(ctx, deploymentName, namespace)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get new deployment %s: %w", deploymentName, err)
+		}
+	}
+	err = r.updateDeployment(ctx, current, desired)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update existing deployment: %w", err)
+	}
+	_, current, err = r.currentControllerDeployment(ctx, deploymentName, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get existing deployment: %w", err)
+	}
+	return current, nil
+}
+
+func desiredDeployment(name, namespace, image string, vpcID, clusterName string, controller *albo.AWSLoadBalancerController, sa *corev1.ServiceAccount) *appsv1.Deployment {
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					appLabelName:    appName,
+					appInstanceName: controller.Name,
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						appLabelName:    appName,
+						appInstanceName: controller.Name,
+					},
+				},
+				Spec: corev1.PodSpec{
+
+					Containers: []corev1.Container{
+						{
+							Name:  "controller",
+							Image: image,
+							Args:  desiredContainerArgs(controller, clusterName, vpcID),
+						},
+					},
+					ServiceAccountName: sa.Name,
+				},
+			},
+		},
+	}
+	if controller.Spec.Config != nil && controller.Spec.Config.Replicas != 0 {
+		d.Spec.Replicas = pointer.Int32(controller.Spec.Config.Replicas)
+	}
+	return d
+}
+
+func desiredContainerArgs(controller *albo.AWSLoadBalancerController, clusterName, vpcID string) []string {
+	var args []string
+	args = append(args, fmt.Sprintf("--aws-vpc-id=%s", vpcID))
```

Comment:
I think we need to add [`hostNetwork` arg](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/7baec176ddbaed9d23f92ca9f77a658eed3cf3a3/helm/aws-load-balancer-controller/values.yaml#L201) as well since it will default to `false`

Reply:
Maybe even the `dnsPolicy` field.

Reply:
Those are values for the deployment spec

https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/7baec176ddbaed9d23f92ca9f77a658eed3cf3a3/helm/aws-load-balancer-controller/templates/deployment.yaml#L46

https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/7baec176ddbaed9d23f92ca9f77a658eed3cf3a3/helm/aws-load-balancer-controller/templates/deployment.yaml#L50

Reply:
Got it, so doesn't it need to be added under our desired deployment?

Reply:
No it's only for clusters where the pods cannot be given AWS credentials that one would need something like that. When the controller runs on the host network it can retrieve credentials from the [AWS metadata endpoint](https://stackoverflow.com/questions/41024362/aws-retrieving-security-credentials-from-instance-metadata). But we don't do that so it's not required.